test_agent/rust/src/utils.rs

```diff
@@ -0,0 +1,528 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * SPDX-FileType: SOURCE
+ * SPDX-FileCopyrightText: 2023 General Motors GTO LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use log::{debug, error};
+use serde::{Deserialize, Deserializer};
+use serde_json::Value;
+use up_rust::{
+    Data, UAttributes, UAuthority, UCode, UEntity, UMessage, UMessageType, UPayload,
+    UPayloadFormat, UPriority, UResource, UStatus, UUri, UUID,
+};
+
+use protobuf::{Enum, MessageField, SpecialFields};
+
+pub fn convert_json_to_jsonstring<T: serde::Serialize>(value: &T) -> String {
+    if let Ok(json_string) = serde_json::to_string(value) {
+        json_string
+    } else {
+        // Handle the error
+        error!("Error: Failed to convert to JSON string");
+        "None".to_owned()
+    }
+}
+
+pub fn get_ustatus_code(u_status: &UStatus) -> u32 {
+    match u_status.get_code() {
+        UCode::OK => 0,
+        UCode::INTERNAL => 13,
+        UCode::ABORTED => 10,
+        UCode::ALREADY_EXISTS => 6,
+        UCode::CANCELLED => 1,
+        UCode::DATA_LOSS => 15,
+        UCode::DEADLINE_EXCEEDED => 4,
+        UCode::FAILED_PRECONDITION => 9,
+        UCode::INVALID_ARGUMENT => 3,
+        UCode::NOT_FOUND => 5,
+        UCode::OUT_OF_RANGE => 11,
+        UCode::PERMISSION_DENIED => 7,
+        UCode::RESOURCE_EXHAUSTED => 8,
+        UCode::UNAUTHENTICATED => 16,
+        UCode::UNAVAILABLE => 14,
+        UCode::UNIMPLEMENTED => 12,
+        UCode::UNKNOWN => 2,
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct WrapperUUri(pub UUri);
+impl<'de> Deserialize<'de> for WrapperUUri {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let value: Value = Deserialize::deserialize(deserializer)?;
+        let authority = if let Ok(authority) = parse_uauthority(&value) {
+            authority
+        } else {
+            let err_msg = "Error parsing authority: ".to_string();
+            error!("{}", err_msg);
+            UAuthority::default()
+        };
+
+        let resource = parse_uresource(&value);
+        let entity = parse_uentity(&value);
+
+        let uuri = if authority.get_name().is_none() && authority.number.is_none() {
+            UUri {
+                entity: MessageField(Some(Box::new(entity))),
```

Comment:
Prefer using the Into trait to reduce line noise:
```suggestion
                entity: Some(entity).into(),
```
Same suggestion for the rest of these lines.